### PR TITLE
Make PacketCaptureAPI optional by moving it to a separate controller

### PR DESCRIPTION
**Pull request by vara2504**
Migrate Packet capture into it own controller.
Design doc: https://docs.google.com/document/d/1bEU1mfjqyR4VBOHiezHBl_mbqhgP2JKFX5d2GgFMe8M/edit#heading=h.4gb7cyubsdxh

calico-private PR:
https://github.com/tigera/calico-private/pull/7544
```
apiVersion: operator.tigera.io/v1
kind: PacketCaptureAPI
metadata:
  creationTimestamp: "2024-05-03T20:59:06Z"
  generation: 1
  name: tigera-secure
  resourceVersion: "290128"
  uid: 1e100aba-28aa-4826-95c1-d238e97d4aa5
status:
  conditions:
  - lastTransitionTime: "2024-05-03T21:00:08Z"
    message: All Objects Available
    observedGeneration: 1
    reason: AllObjectsAvailable
    status: "False"
    type: Progressing
  - lastTransitionTime: "2024-05-03T21:00:08Z"
    message: All Objects Available
    observedGeneration: 1
    reason: AllObjectsAvailable
    status: "False"
    type: Degraded
  - lastTransitionTime: "2024-05-03T21:00:08Z"
    message: All objects available
    observedGeneration: 1
    reason: AllObjectsAvailable
    status: "True"
    type: Ready
  state: Ready

```

```
vara@vara:~/bzprofiles/Clusters/tc_mgmt$ k get tigerastatus packet-capture
NAME             AVAILABLE   PROGRESSING   DEGRADED   SINCE
packet-capture   True        False         False      35m

```

## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

File: pkg/controller/packetcapture/packetcapture_controller.go [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+	managementCluster, err := utils.GetManagementCluster(ctx, r.client)
+	if err != nil {
+		r.status.SetDegraded(operatorv1.ResourceReadError, "Error reading ManagementCluster", err, reqLogger)
+		return reconcile.Result{}, err
+	}
+
+	// Packet capture is disabled in multi tenant management cluster
+	if r.multiTenant && managementCluster != nil {
+		return reconcile.Result{}, err
+	}
```
Comment (vara2504) [inline pkg/controller/packetcapture/packetcapture_controller.go:200]: Should i set the status degraded if we encounter "packet capture" in multi-tenant management cluster? @caseydavenport

Comment (vara2504) [reply pkg/controller/packetcapture/packetcapture_controller.go:206]: Updated the code: The packet capture  would be in a degraded status if it is found inside a multitenant management cluster.

File: pkg/controller/packetcapture/packetcapture_controller.go [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+		if errors.IsNotFound(err) {
+			r.status.SetDegraded(operatorv1.ResourceNotFound, "Installation not found", err, reqLogger)
+			return reconcile.Result{}, err
+		}
+		r.status.SetDegraded(operatorv1.ResourceReadError, "Error querying installation", err, reqLogger)
+		return reconcile.Result{}, err
+	}
+
+	if variant != operatorv1.TigeraSecureEnterprise {
+		r.status.SetDegraded(operatorv1.ResourceNotReady, fmt.Sprintf("Waiting for network to be %s", operatorv1.TigeraSecureEnterprise), nil, reqLogger)
```
Comment (caseydavenport) [inline pkg/controller/packetcapture/packetcapture_controller.go:192]: nit, but this should be "Waiting for Installation variant to be" (network is outdated terminology)

File: pkg/controller/packetcapture/packetcapture_controller.go [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+
+	managementCluster, err := utils.GetManagementCluster(ctx, r.client)
+	if err != nil {
+		r.status.SetDegraded(operatorv1.ResourceReadError, "Error reading ManagementCluster", err, reqLogger)
+		return reconcile.Result{}, err
+	}
+
+	// Packet capture is disabled in multi tenant management cluster
+	if r.multiTenant && managementCluster != nil {
+		r.status.SetDegraded(operatorv1.ResourceNotReady, "Packet capture is disabled in the multitenant management cluster.", err, reqLogger)
```
Comment (caseydavenport) [inline pkg/controller/packetcapture/packetcapture_controller.go:204]: should say "Packet capture is not supported on multi-tenant management clusters"

File: pkg/controller/packetcapture/packetcapture_controller.go [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+func newReconciler(mgr manager.Manager, opts options.AddOptions, licenseAPIReady *utils.ReadyFlag, tierWatchReady *utils.ReadyFlag) reconcile.Reconciler {
+	r := &ReconcilePacketCapture{
+		client:              mgr.GetClient(),
+		scheme:              mgr.GetScheme(),
+		provider:            opts.DetectedProvider,
+		enterpriseCRDsExist: opts.EnterpriseCRDExists,
+		status:              status.New(mgr.GetClient(), ResourceName, opts.KubernetesVersion),
+		clusterDomain:       opts.ClusterDomain,
+		usePSP:              opts.UsePSP,
+		licenseAPIReady:     licenseAPIReady,
```
Comment (caseydavenport) [inline pkg/controller/packetcapture/packetcapture_controller.go:118]: Do we need the license watch in this controller? Looks like we never query the license.

File: pkg/controller/packetcapture/packetcapture_controller.go [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+		}
+		r.status.SetDegraded(operatorv1.ResourceNotReady, "Error querying allow-tigera tier", err, reqLogger)
+		return reconcile.Result{}, err
+	}
+
+	// Create a component handler to manage the rendered component.
+	handler := utils.NewComponentHandler(log, r.client, r.scheme, packetcaptureapi)
+
+	// Render the desired objects from the CRD and create or update them.
+	reqLogger.V(3).Info("rendering components")
```
Comment (caseydavenport) [inline pkg/controller/packetcapture/packetcapture_controller.go:233]: We can get rid of this log, there are logs in the handler code that tell us this instead.

**Review by caseydavenport — COMMENTED** [review]
Just a few minor comments

File: pkg/controller/packetcapture/packetcapture_controller.go (hunk 4 above)
Comment (vara2504) [reply pkg/controller/packetcapture/packetcapture_controller.go:118]: @caseydavenport  I will remove this. Just wondering if we need a license check for packet capture. Initially, it was part of [apiserver_controller.go](https://github.com/tigera/operator/blob/1565f374e78473c889726665b5c21ba62be88549/pkg/controller/apiserver/apiserver_controller.go#L408), and it didn't have any license check at that time.

**Review by caseydavenport — APPROVED** [review]

**Pull request merged**